components/compositing/compositor.rs

```diff
@@ -831,16 +822,39 @@ impl<Window: WindowMethods> IOCompositor<Window> {
                 self.on_pinch_zoom_window_event(magnification);
             }
 
-            WindowEvent::Navigation(direction) => {
-                self.on_navigation_window_event(direction);
+            WindowEvent::Navigation(top_level_browsing_context_id, direction) => {
+                let direction = match direction {
+                    windowing::WindowNavigateMsg::Forward => TraversalDirection::Forward(1),
+                    windowing::WindowNavigateMsg::Back => TraversalDirection::Back(1),
+                };
+                let msg = ConstellationMsg::TraverseHistory(top_level_browsing_context_id, direction);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending navigation to constellation failed ({}).", e);
+                }
             }
 
             WindowEvent::TouchpadPressure(cursor, pressure, stage) => {
                 self.on_touchpad_pressure_event(cursor, pressure, stage);
             }
 
             WindowEvent::KeyEvent(ch, key, state, modifiers) => {
-                self.on_key_event(ch, key, state, modifiers);
+                // FIXME(paul): this should be a message sent from embedder
+                // Steal a few key events for webrender debug options.
+                if modifiers.contains(CONTROL) && state == KeyState::Pressed {
```

Comment:
It would be really nice if we didn't need this sort of hand-crafted code in servo itself, can we find a way to move it into the embedding app?